### PR TITLE
TSV-497: Tax Account Router now returns Agent as AccountType for Agent users.

### PR DESCRIPTION
**Pull request by cellis-ee**

### CODE REVIEW

File: app/controllers/internal/AccountTypeController.scala [hunk 1]
```diff
@@ -63,12 +68,25 @@ trait AccountTypeController extends FrontendController with Actions {
...
+        val auditContext = createAuditContext()
+        ruleEngine.matchRulesForLocation(ruleContext, auditContext) map { location =>
+          val accountType: AccountType = accountTypeBasedOnLocation(location)
+          Ok(Json.toJson(AccountTypeResponse(accountType)))
+        }
+    }.recover {
+      case e =>
+        e.printStackTrace()
+        logger.error(s"Failed to get user details due to downstream failure.", e)
+        InternalServerError("Failed to get user details due to downstream failure.")
```
Comment (cellis-ee) [inline app/controllers/internal/AccountTypeController.scala]: Not sure about the word "failure", but also not overly concerned. @dclifton-hmrc thoughts?

File: it/router/AccountTypeForCredIdFeature.scala [hunk 2]
```diff
@@ -48,7 +50,7 @@ class AccountTypeForCredIdFeature extends StubbedFeatureSpec with CommonStubs {
       verify(getRequestedFor(urlEqualTo("/auth/enrolments-uri")))
 
       And("user's details should not be fetched from User Details")
-      verify(0, getRequestedFor(urlEqualTo("/user-details-uri")))
+      verify(1, getRequestedFor(urlEqualTo("/user-details-uri")))
```
Comment (cellis-ee) [inline it/router/AccountTypeForCredIdFeature.scala]: I think '1' is the default here, if you want to get rid of a few characters ;D

File: it/router/AccountTypeForCredIdFeature.scala [hunk 3]
```diff
@@ -48,7 +50,7 @@ class AccountTypeForCredIdFeature extends StubbedFeatureSpec with CommonStubs {
       verify(getRequestedFor(urlEqualTo("/auth/enrolments-uri")))
 
       And("user's details should not be fetched from User Details")
```
Comment (dclifton-hmrc) [inline it/router/AccountTypeForCredIdFeature.scala]: all the comments are now wrong `user's details should be fetched from User Details` - remove the `not`

File: it/router/AccountTypeForCredIdFeature.scala [hunk 4]
```diff
@@ -327,27 +338,51 @@ class AccountTypeForCredIdFeature extends StubbedFeatureSpec with CommonStubs {
...
+      And("the authority object should be fetched once for AuthenticatedBy")
+      verify(getRequestedFor(urlEqualTo(s"/auth/gg/$credId")))
+
+      And("user's enrolments should not be fetched from Auth")
+      verify(0, getRequestedFor(urlEqualTo("/auth/enrolments-uri")))
+
+      And("user's details should be fetched from User Details")
+      verify(getRequestedFor(urlEqualTo("/user-details-uri")))
+
+      And("Sa micro service should not be invoked")
```
Comment (dclifton-hmrc) [inline it/router/AccountTypeForCredIdFeature.scala]: `SA`

File: it/router/AccountTypeForCredIdFeature.scala [hunk 5]
```diff
@@ -327,27 +338,51 @@ class AccountTypeForCredIdFeature extends StubbedFeatureSpec with CommonStubs {
       verify(0, getRequestedFor(urlMatching("/sa/individual/.[^\\/]+/return/last")))
     }
 
-    scenario("a user logged in through GG and has no sa and no business enrolment and no inactive enrolments and affinity group not available should get Organisation as account type") {
+    scenario("a user logged in through GG and affinity group not available should get 500 response") {
```
Comment (dclifton-hmrc) [inline it/router/AccountTypeForCredIdFeature.scala]: it feels like some parts of the old test `a user logged in through GG and has no sa and no business enrolment and no inactive enrolments and affinity group not available should get Organisation as account type` are still valid.  I think we need a `a user logged in through GG and has no sa and no business enrolment and no inactive enrolments...` test.

File: app/controllers/internal/AccountTypeController.scala [hunk 6]
```diff
@@ -63,12 +68,25 @@ trait AccountTypeController extends FrontendController with Actions {
...
+        Future.successful(Ok(Json.toJson(AccountTypeResponse(AccountType.Agent))))
+      case _ =>
+        val auditContext = createAuditContext()
+        ruleEngine.matchRulesForLocation(ruleContext, auditContext) map { location =>
+          val accountType: AccountType = accountTypeBasedOnLocation(location)
+          Ok(Json.toJson(AccountTypeResponse(accountType)))
+        }
+    }.recover {
+      case e =>
+        e.printStackTrace()
```
Comment (dclifton-hmrc) [inline app/controllers/internal/AccountTypeController.scala]: don't think we want this in the production code

File: app/controllers/internal/AccountTypeController.scala [hunk 7]
```diff
@@ -63,12 +68,25 @@ trait AccountTypeController extends FrontendController with Actions {
...
+        logger.error(s"Failed to get user details due to downstream failure.", e)
```
Comment (dclifton-hmrc) [inline app/controllers/internal/AccountTypeController.scala]: no need for `s` there's no interpolation.

File: app/controllers/internal/AccountTypeController.scala (hunk 1 above)
Comment (dclifton-hmrc) [inline app/controllers/internal/AccountTypeController.scala]: perhaps: 'Unable to get user details due to downstream failure'

File: it/router/AccountTypeForCredIdFeature.scala [hunk 8]
```diff
@@ -81,10 +81,10 @@ class AccountTypeForCredIdFeature extends StubbedFeatureSpec with CommonStubs {
       And("user's enrolments should be fetched from Auth")
       verify(getRequestedFor(urlEqualTo("/auth/enrolments-uri")))
 
-      And("user's details should not be fetched from User Details")
-      verify(1, getRequestedFor(urlEqualTo("/user-details-uri")))
+      And("user's details should be fetched from User Details")
+      verify(getRequestedFor(urlEqualTo("/user-details-uri")))
 
-      And("sa returns should be fetched from Sa micro service")
+      And("sa returns should be fetched from SA micro service")
```
Comment (cellis-ee) [inline it/router/AccountTypeForCredIdFeature.scala]: while you're changing `Sa` -> `SA`, you could probably also change `sa` -> `SA` (as in `sa returns`)

**Pull request merged**